File: server/sonar-web/src/main/webapp/WEB-INF/lib/database_version.rb
#
# SonarQube, open source software quality management tool.
# Copyright (C) 2008-2014 SonarSource
# mailto:contact AT sonarsource DOT com
#
# SonarQube is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# SonarQube is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#

class DatabaseVersion

  class DeprecatedSchemaInfo < ActiveRecord::Base
    set_table_name 'schema_info'
  end

  def self.current_version
    begin
      result=ActiveRecord::Migrator.current_version
    rescue
      result=0
    end

    if result==0
      begin
        result=DeprecatedSchemaInfo.find(:first).version
      rescue
      end
    end
    result
  end

  def self.target_version
    files = Dir["#{migrations_path}/[0-9]*_*.rb"].sort
    files.last.scan(/([0-9]+)_[_a-z0-9]*.rb/).first[0].to_i
  end

  def self.migrations_path
    File.dirname(__FILE__).to_s + "/../db/migrate/"
  end

  $uptodate = false

  def self.uptodate?
    unless $uptodate
      $uptodate = (current_version>=target_version)
    end
    $uptodate
  end

  def self.upgrade
    ActiveRecord::Migrator.migrate(migrations_path)
  end

  def self.upgrade_and_start
    ActiveRecord::Migrator.migrate(migrations_path)
    Java::OrgSonarServerPlatform::Platform.getInstance().doStart()
    load_java_web_services
  end

  def self.load_java_web_services
    ActionController::Routing::Routes.add_java_ws_routes
  end

  def self.automatic_setup
    if current_version<=0
      upgrade_and_start()
    else
      load_java_web_services
    end
    uptodate?
  end

  def self.connected?
    ActiveRecord::Base.connected?
  end

  def self.dialect
    ::Java::OrgSonarServerUi::JRubyFacade.getInstance().getDatabase().getDialect().getActiveRecordDialectCode()
  end

  def self.production?
    @@production ||=
      begin
        dialect()!='.h2.'
      end
  end
end

File: sonar-core/src/main/java/org/sonar/core/persistence/DatabaseVersion.java
/*
 * SonarQube, open source software quality management tool.
 * Copyright (C) 2008-2014 SonarSource
 * mailto:contact AT sonarsource DOT com
 *
 * SonarQube is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * SonarQube is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.core.persistence;

import com.google.common.annotations.VisibleForTesting;
import com.google.common.collect.ImmutableList;
import org.apache.ibatis.session.SqlSession;
import org.sonar.api.BatchSide;
import org.sonar.api.ServerSide;

import java.util.Collections;
import java.util.List;

/**
 * @since 3.0
 */
@BatchSide
@ServerSide
public class DatabaseVersion {

  public static final int LAST_VERSION = 914;

  /**
   * List of all the tables.n
   * This list is hardcoded because we didn't succeed in using java.sql.DatabaseMetaData#getTables() in the same way
   * for all the supported databases, particularly due to Oracle results.
   */
  public static final List<String> TABLES = ImmutableList.of(
    "action_plans",
    "active_dashboards",
    "active_rules",
    "active_rule_parameters",
    "activities",
    "analysis_reports",
    "authors",
    "characteristics",
    "dashboards",
    "dependencies",
    "duplications_index",
    "events",
    "file_sources",
    "graphs",
    "groups",
    "groups_users",
    "group_roles",
    "issues",
    "issue_changes",
    "issue_filters",
    "issue_filter_favourites",
    "loaded_templates",
    "manual_measures",
    "measure_filters",
    "measure_filter_favourites",
    "metrics",
    "notifications",
    "permission_templates",
    "perm_templates_users",
    "perm_templates_groups",
    "quality_gates",
    "quality_gate_conditions",
    "projects",
    "project_links",
    "project_measures",
    "project_qprofiles",
    "properties",
    "resource_index",
    "rules",
    "rules_parameters",
    "rules_profiles",
    "semaphores",
    "schema_migrations",
    "snapshots",
    "users",
    "user_roles",
    "widgets",
    "widget_properties"
    );
  private MyBatis mybatis;

  public DatabaseVersion(MyBatis mybatis) {
    this.mybatis = mybatis;
  }

  @VisibleForTesting
  static Status getStatus(Integer currentVersion, int lastVersion) {
    Status status = Status.FRESH_INSTALL;
    if (currentVersion != null) {
      if (currentVersion == lastVersion) {
        status = Status.UP_TO_DATE;
      } else if (currentVersion > lastVersion) {
        status = Status.REQUIRES_DOWNGRADE;
      } else {
        status = Status.REQUIRES_UPGRADE;
      }
    }
    return status;
  }

  public Status getStatus() {
    return getStatus(getVersion(), LAST_VERSION);
  }

  public Integer getVersion() {
    SqlSession session = mybatis.openSession(false);
    try {
      List<Integer> versions = session.getMapper(SchemaMigrationMapper.class).selectVersions();
      if (!versions.isEmpty()) {
        Collections.sort(versions);
        return versions.get(versions.size() - 1);
      }
      return null;
    } catch (RuntimeException e) {
      // The table SCHEMA_MIGRATIONS does not exist.
      // Ignore this exception -> it will be created by Ruby on Rails migrations.
      return null;

    } finally {
      MyBatis.closeQuietly(session);
    }
  }

  public static enum Status {
    UP_TO_DATE, REQUIRES_UPGRADE, REQUIRES_DOWNGRADE, FRESH_INSTALL
  }
}
